lib/libc/newlib/libc-hooks.c

```diff
@@ -298,17 +299,130 @@ void *_sbrk(intptr_t count)
 }
 __weak FUNC_ALIAS(_sbrk, sbrk, void *);
 
-static LIBC_DATA SYS_MUTEX_DEFINE(heap_mutex);
+#ifdef CONFIG_MULTITHREADING
+/*
+ * Newlib Retargetable Locking Interface Implementation
+ *
+ * When multithreading is enabled, the newlib retargetable locking interface is
+ * defined below to override the default void implementation and provide the
+ * Zephyr-side locks.
+ */
+
+/* Static locks */
+struct sys_mutex __lock___sinit_recursive_mutex;
+struct sys_mutex __lock___sfp_recursive_mutex;
+struct sys_mutex __lock___atexit_recursive_mutex;
+struct sys_mutex __lock___malloc_recursive_mutex;
+struct sys_mutex __lock___env_recursive_mutex;
+struct sys_sem __lock___at_quick_exit_mutex;
+struct sys_sem __lock___tz_mutex;
+struct sys_sem __lock___dd_hash_mutex;
+struct sys_sem __lock___arc4random_mutex;
+
+/* Initialise all static locks after boot */
+static int newlib_locks_prepare(const struct device *unused)
+{
+	ARG_UNUSED(unused);
+
+	/* Initialise recursive locks */
+	sys_mutex_init(&__lock___sinit_recursive_mutex);
+	sys_mutex_init(&__lock___sfp_recursive_mutex);
+	sys_mutex_init(&__lock___atexit_recursive_mutex);
+	sys_mutex_init(&__lock___malloc_recursive_mutex);
+	sys_mutex_init(&__lock___env_recursive_mutex);
+
+	/* Initialise non-recursive locks */
+	sys_sem_init(&__lock___at_quick_exit_mutex, 1, 1);
+	sys_sem_init(&__lock___tz_mutex, 1, 1);
+	sys_sem_init(&__lock___dd_hash_mutex, 1, 1);
+	sys_sem_init(&__lock___arc4random_mutex, 1, 1);
+
+	return 0;
+}
+
+SYS_INIT(newlib_locks_prepare, POST_KERNEL,
+	 CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
+
+/* Create a new dynamic non-recursive lock */
+void __retarget_lock_init(_LOCK_T *lock)
+{
+	__ASSERT_NO_MSG(lock != NULL);
+
+	/* Allocate semaphore object */
+	*lock = malloc(sizeof(struct sys_sem));
+	__ASSERT(*lock != NULL, "non-recursive lock allocation failed");
+
+	sys_sem_init((struct sys_sem *)*lock, 1, 1);
+}
+
+/* Create a new dynamic recursive lock */
+void __retarget_lock_init_recursive(_LOCK_T *lock)
+{
+	__ASSERT_NO_MSG(lock != NULL);
+
+	/* Allocate mutex object */
+	*lock = malloc(sizeof(struct sys_mutex));
+	__ASSERT(*lock != NULL, "recursive lock allocation failed");
+
+	sys_mutex_init((struct sys_mutex *)*lock);
```

Comment:
So, unlike `sys_sem`, `sys_mutex` does not provide the userspace implementation (i.e. futex) and `sys_mutex_init` is, for some reason, no-op simply relying on the build system to provide the kernel-side object only for the statically allocated `sys_sem`s.

This dynamically allocated `sys_mutex` should not work when `CONFIG_USERSPACE=y` because there is no corresponding `k_mutex` kernel object:

https://github.com/zephyrproject-rtos/zephyr/blob/21d1ad3762302b3e461953df59430c77e0709274/lib/os/mutex.c#L33-L42

https://github.com/zephyrproject-rtos/zephyr/blob/21d1ad3762302b3e461953df59430c77e0709274/lib/os/mutex.c#L12-L22

I would say the fact that `sys_mutex_init` is a no-op, when there is no futex, is a bug.

@andyross Any comments on this?

Reply:
After a more in-depth analysis on the current Zephyr userspace implementation, even the `k_futex` used internally by `sys_mutex` will have problems when there is a contention because the `k_futex` relies on the `z_futex_data` kernel object.

Ideally, it would be possible to dynamically allocate these objects through `k_object_alloc`, but the current implementation prohibits that -- the `Lives in user memory` rationale given here is incorrect because these "kernel" objects are supposed to be the protected kernel-side objects for the corresponding user objects.
https://github.com/zephyrproject-rtos/zephyr/blob/d6c9d40ee062f7dd1f8edae461e21c8fe88824db/kernel/userspace.c#L347-L354

FWIW, I think the current userspace `sys_sem` and `sys_mutex` implementations require some major rework to provide kind of abstraction that it was supposed to provide. Alternatively, we could simply declare that dynamic allocation of these objects are simply not allowed given the nature of Zephyr "RT"OS, but that sounds like a major limitation for more advanced applications.

@andyross Your insight would be greatly appreciated.

Reply:
Related issues:
* #25497 `k_futex` unnecessarily relies on the kernel object management
* #15138 `sys_mutex` does not use `k_futex` for userspace